### PR TITLE
Switch to GHA and GHCR for image building and hosting

### LINKED ISSUE #30: IDEA: move image generation to GHA, and image storage to GHCR
The Dockerfiles we build aren't especially complex and Docker – with its various team / org / resource restrictions – isn't my favourite service to use. Given that there are high-quality actions available for building images and that we only use the images in Github, is it worth exploring this? I like the idea of driving everything from an action using a `libproj_version` x `rust_version` matrix.

### PR DESCRIPTION
**Pull request by urschrei**
Closes #30

### CODE REVIEW

**Comment by urschrei** [comment]
@michaelkirk This isn't going _terribly_, but I have two questions:

1. I'm confused about how the local build commands work, given that three of them are multi-stage builds which pull the latest `libproj-builder` from Docker Hub in order to build. This obviously works locally, but I don't understand _how_.
2. What are we going to do about testing, given that we don't have e.g. `geo` available on the test runners?

**Comment by urschrei** [comment]
The current setup works as follows:

For each Rust version in the matrix
1. A `libproj-builder` image is built using the specific `libproj` version, tagged, and pushed to ghcr.
2. Upon success of that job, the three dependent images are built, tagged, and pushed to ghcr

We have some images for inspection / testing:

https://github.com/orgs/georust/packages?repo_name=docker-images

We need to figure out what testing looks like for these images: and the current official Docker action always pushes upon a successful build, so ideally the testing needs to be defined in the respective Dockerfiles, somehow.

Things like tagging something 'latest' can be done on push of a tag.

**Comment by michaelkirk** [comment]
> I'm confused about how the local build commands work, given that three of them are multi-stage builds which pull the latest libproj-builder from Docker Hub in order to build. This obviously works locally, but I don't understand how.

It seems like you've since figured this out - but just to be clear: The build process was not pulling the earlier stages from Docker Hub. They were building and tagging them locally. Once built, the images are available from the local registry.  Only after successfully running the tests against the built containers would I push anything to Docker Hub.

> What are we going to do about testing, given that we don't have e.g. geo available on the test runners?

With the current system, I can run tests against the new containers before pushing them to be reasonably confident that the work. To maintain that kind of testing, my assumption is we'd clone down the necessary repos (georust/geo, georust/proj) as part of this new CI job, and then run the tests similarly to how they're run now before the containers are pushed to ghcr.io

I know this is still WIP, but it'd be helpful to update the README to clarify what the new process for getting up-to-date containers in, e.g., georust/geo CI looks like. And relatedly, is the tagging scheme going to change from what's in the current container build scripts?

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:33]: Multi arch could be nice, but it's not something that we ever supported before AFAIK. I'd encourage you not to make multiarch a requirement for getting this merged.

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:55]: Ah - I see the tag scheme here. Looks like it's the same as before.

Comment (urschrei) [reply .github/workflows/imagebuild.yml:55]: Yep, it's the same!

**Comment by urschrei** [comment]
> With the current system, I can run tests against the new containers before pushing them to be reasonably confident that the work. To maintain that kind of testing, my assumption is we'd clone down the necessary repos (georust/geo, georust/proj) as part of this new CI job, and then run the tests similarly to how they're run now before the containers are pushed to ghcr.io

Oh that's a good idea. No clue how I'm gonna make that work in the test step, but I'll see how I fare.

>  know this is still WIP, but it'd be helpful to update the README to clarify what the new process for getting up-to-date containers in, e.g., georust/geo CI looks like. And relatedly, is the tagging scheme going to change from what's in the current container build scripts?

Yeah, I just haven't gotten around to it yet – no point in doing it til we know that this works the way we want.

Adding new Rust versions involves updating two matrices in the workflow with the version you want, and updating the `libproj` version involves updating the env var. If the containers build and upload, the action passes and they're immediately available to us using the same naming scheme we've been using, e.g. `ghcr.io/georust/libproj-builder:proj-9.4.0-rust-1.75`. We _could_ pull in Rust and `libproj` versions from text files at action run time, but it'll add some complexity.

**Comment by urschrei** [comment]
Tests are passing in the built images, and README is updated.

Two things:

1. Are we comfortable pushing new versions when tests pass, or would we prefer a "manual" publishing step, e.g. when we push a new git tag? Successfully built images can be uploaded as Github artifacts, and the "publish" step can download them and push them to the registry. It's more work to configure, but should work fine.

2. I simplified the tests in `proj-ci-without-system-proj` because it's going to be pain to run three times to simulate bundled / non-bundled `libproj`. We currently just test the `bundled-proj` feature, which is the point of the image anyway. Tests for the other two images have been reproduced from the Makefile.

**Comment by urschrei** [comment]
(I haven't added the Rust versions 1.70 - 1.75 yet because it will quintuple the test-build cycle)

**Comment by michaelkirk** [comment]
Screenshot 2024-04-03 at 16 03 38

It seems like all the container builds are running the same set of tests. e.g. the geo-ci container is running the proj tests, no the georust/geo tests.

Probably you can fix this in the workflow file. Another approach would be, rather than having 3 separate "set up RUN_CMD" steps, to instead have a `test_$CONTAINER_NAME.sh` script in the repository.

Either would be fine.

Also, it seems like we're not longer running the `_PROJ_SYS_TEST_EXPECT_BUILD_FROM_SRC` tests from the deleted Makefile. Is that intentional?

File: README.md
Comment (michaelkirk) [inline README.md:33]: The containers are published regardless of if the PR is merged, right? Ideally it'd be nice to not publish until merged into `main` but it seems non-trivial - building the three downstream containers across different jobs is nice for parallelism, but it means we can't rely on the local registry.

I can live with things as you have them now - but if you can think of an easy way to hold off on publishing until an merge into main, that'd be cool. Otherwise we can leave it for future work.

File: README.md
Comment (michaelkirk) [inline README.md:51]: less instructions is usually a good sign. =)

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:23]: I wonder if we actually need to specify rust_version for the libprojbuilder - it's not even using rust, right? It's fine to leave it this way for now, since that's how it's always been, but worth considering for future simplification. It'd speed up the build a bit and avoid having to track supported versions in two places.

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:80]: We're no longer running proj-sys tests. Was that intentionally dropped?

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:79]: I think you could add `_PROJ_SYS_TEST_EXPECT_BUILD_FROM_SRC=0` except for `bundled_proj` which should `_PROJ_SYS_TEST_EXPECT_BUILD_FROM_SRC=1`.

File: dockerfiles/proj-ci-without-system-proj
Comment (michaelkirk) [inline dockerfiles/proj-ci-without-system-proj:21]: I'm curious why the package dependencies have changed in all the images. They look to be deps for building proj, but why did it change?

File: .github/workflows/imagebuild.yml
Comment (michaelkirk) [inline .github/workflows/imagebuild.yml:82]: This should include all the supported versions before we're "done" here, right?

This is my only blocking concern - everything else is just informational.

**Review by michaelkirk — APPROVED** [review]

File: .github/workflows/imagebuild.yml
Comment (urschrei) [reply .github/workflows/imagebuild.yml:82]: Yep, this is purely for dev purposes until before I merge because it'll multiply the build time.

File: dockerfiles/proj-ci-without-system-proj
Comment (urschrei) [reply dockerfiles/proj-ci-without-system-proj:21]: I was seeing some build failures last night – let me try removing it again.

File: .github/workflows/imagebuild.yml
Comment (urschrei) [reply .github/workflows/imagebuild.yml:23]: Oooh. That's a good point.

File: dockerfiles/proj-ci-without-system-proj
Comment (urschrei) [reply dockerfiles/proj-ci-without-system-proj]: Passing now. I hate gremlins.

File: .github/workflows/imagebuild.yml
Comment (urschrei) [reply .github/workflows/imagebuild.yml:25]: The `libproj-builder` Docker image does expect a Rust version:

`ARG RUST_VERSION`

Unless I'm misunderstanding something about how the multi-stage builds work?

**Comment by michaelkirk** [comment]
indeed it does - but I don’t think it’s used for anything. We’d also have to update the dependent stages to reference the new tag for libprojbuilder who can wouldn’t include the rust portion of the tag id On Apr 4, 2024, at 11:51, Stephan Hügel ***@***.***> wrote:﻿
@urschrei commented on this pull request.



In .github/workflows/imagebuild.yml:
> +      - trying
+  pull_request:
+  merge_group:
+
+env:
+  LIBPROJ_VERSION: 9.4.0
+  MAIN_IMAGE_NAME: libproj-builder
+
+jobs:
+  build_main_image:
+    name: Build the main image for geo and proj
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        rust_version: [1.77]

The libproj-builder Docker image does expect a Rust version:
ARG RUST_VERSION
Unless I'm misunderstanding something about how the multi-stage builds work?

**Comment by urschrei** [comment]
> indeed it does - but I don’t think it’s used for anything. We’d also have to update the dependent stages to reference the new tag for libprojbuilder who can wouldn’t include the rust portion of the tag id On Apr 4, 2024, at 11:51, Stephan Hügel ***@***.***> wrote:﻿ @urschrei commented on this pull request. In .github/workflows/imagebuild.yml:
> + - trying
> + pull_request: + merge_group: + +env: + LIBPROJ_VERSION: 9.4.0 + MAIN_IMAGE_NAME: libproj-builder + +jobs: + build_main_image: + name: Build the main image for geo and proj + runs-on: ubuntu-latest + + strategy: + matrix: + rust_version: [1.77] The libproj-builder Docker image does expect a Rust version: ARG RUST_VERSION Unless I'm misunderstanding something about how the multi-stage builds work? —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

Happy to leave this as a follow-up action.

File: README.md
Comment (urschrei) [reply README.md:33]: GHA now has the ability to trigger a job if a PR is merged.
In order to take advantage of that, we'd have to add a few new moving parts:

1. Upload built `libproj-builder` x rust versions (for now) images as artifacts
2. Upload each sub-image x rust version artifact
3. On successful merge, download all the above again, then push to the registry.

It can be done, but it'll be easier if we can implement your no-Rust-for-main-image idea first. It will also mean duplicating the rust version and image name matrices because matrices can't be shared between jobs. That strikes me as error-prone, so we might have to explore building matrices from JSON files at job run time.

**Comment by urschrei** [comment]
Images for 1.70 (MSRV), 1.76, and 1.77 are now in the registry.

File: dockerfiles/proj-ci
Comment (michaelkirk) [inline dockerfiles/proj-ci:15]: Why this new dependency?

File: dockerfiles/libproj-builder
Comment (michaelkirk) [inline dockerfiles/libproj-builder:17]: Why this new dependency?

File: dockerfiles/geo-ci
Comment (michaelkirk) [inline dockerfiles/geo-ci:24]: Why this new dependency?

File: README.md
Comment (michaelkirk) [reply README.md:33]: My first though was to keep the publish step in the existing workflow, but gate it on a `if branch == main` check.

I'm happy to follow up with that later though.

**Review by michaelkirk — APPROVED** [review]
just minor notes. Feel free to merge this.

File: README.md
Comment (urschrei) [reply README.md:33]: Oh, that might be a lot easier…

**Pull request merged**
